### PR TITLE
Hardhat refactor 👷

### LINKED ISSUE #416: Re-writing the Truffle migrations into a deployment script that can be run using hardhat-deploy

### LINKED ISSUE #417: Refactor test suite to work with hardhat + Waffle
Checklist:

- [x] Make sure that the eth-gas-reporter hardhat plugin can perform gas profiling
- [x] Make sure that coverage reporting works with coveralls

NOT in scope:

- Adding TS features to tests
- Adding TS bindings to tests

We'll leave the things listed under not in scope to be addressed separately.

### LINKED ISSUE #441: replace scripts/parseArtifacts.js with hardhat-abi-exporter
Currently the `scripts/parseArtifacts.js` file is used to extract ABIs from truffle artifacts. Since this is now deprecated let's replace this functionality with the `hardhat-abi-exporter` plugin. 

https://hardhat.org/plugins/hardhat-abi-exporter.html

### PR DESCRIPTION
**Pull request by kyriediculous**
**What does this pull request do? Explain your changes. (required)**
Hardhat refactor 👷 

**Specific updates (required)**

**general**
- Added hardhat-deploy plugin
- Create custom contract deployer using hardhat-deploy so it works with out proxies
- Set `allowJs: true` in `tsconfig`
- Switch to hardhat-web3
- Use ES6 import instead of `require` for module imports
- Fix test running
- Deployment script for polls https://github.com/livepeer/protocol/pull/435
- Cleanup dependencies, remove other legacy truffle related scripts https://github.com/livepeer/protocol/pull/444
- Use hardhat-abi-exporter plugin , this exports all contract ABIs into an `/abi` folder 
- Fix ESlint for both .js and .ts files https://github.com/livepeer/protocol/pull/445

**unit tests**
- Fix controller unit tests https://github.com/livepeer/protocol/pull/421
- Fix governor unit tests https://github.com/livepeer/protocol/pull/422
- Fix faucet unit tests https://github.com/livepeer/protocol/pull/423
- Fix ManagerProxy unit tests https://github.com/livepeer/protocol/pull/424
- Fix MerkleSNapshots unit tests https://github.com/livepeer/protocol/pull/425
- Fix Poll, Poll creator unit tests https://github.com/livepeer/protocol/pull/426
- Fix ServiceRegistry unit tests https://github.com/livepeer/protocol/pull/426
- Fix VariableSupplyToken unit tests https://github.com/livepeer/protocol/pull/426
- Fix BondingManager unit tests https://github.com/livepeer/protocol/pull/427
- Fix RoundsManager unit test https://github.com/livepeer/protocol/pull/430
- Fix solidity unit tests https://github.com/livepeer/protocol/pull/436

**integration tests**
- Fix TicketBroker integration tests https://github.com/livepeer/protocol/pull/428
- Fix Minter integration tests https://github.com/livepeer/protocol/pull/429
- Fix Delegation integration test https://github.com/livepeer/protocol/pull/432
- Fix gas-reports and remaining integration tests https://github.com/livepeer/protocol/pull/433

**How did you test each of these updates (required)**
Ran test suite

**Does this pull request close any open issues?**
Fixes #416 
Fixes #417 
Fixes #441 

**Checklist:**
- [x] README and other documentation updated
- [x] All tests using `npm run test` pass

### CODE REVIEW

File: hardhat.config.ts [hunk 1]
```diff
@@ -2,31 +2,64 @@ import "@nomiclabs/hardhat-ethers";
...
   solidity: {
-    version: "0.5.11",
-    settings: {
-      optimizer: {
-        enabled: true,
-        runs: 200,
-      },
-    },
+    compilers: [
+      { version: "0.5.11",
```
Comment (yondonfu) [inline hardhat.config.ts]: Bracket indentation fix needed.

File: package.json [hunk 2]
```diff
@@ -44,21 +46,25 @@
   },
   "homepage": "https://github.com/livepeer/protocol#readme",
   "devDependencies": {
-    "@nomiclabs/hardhat-ethers": "^2.0.2",
+    "@nomiclabs/hardhat-ethers": "npm:hardhat-deploy-ethers@^0.3.0-beta.10",
```
Comment (yondonfu) [inline package.json:44]: I'm guessing we need something from this particular version? If so, what do we need?

**Review by yondonfu — COMMENTED** [review]
Looking good so far! Looks like a few tests are failing in CI though (I'm sure y'all are aware, but just making note of it).

File: package.json (hunk 2 above)
Comment (kyriediculous) [reply package.json:44]: the hardhat version of ethers has some slight modifications that can be found here https://hardhat.org/plugins/nomiclabs-hardhat-ethers.html#helpers

hardhat-deploy further extends this functionality with helpers such as getting a signer based on an address and some other functions , you can view the full interface extension here: https://www.npmjs.com/package/hardhat-deploy-ethers

This version was automatically installed when installing `hardhat-deploy`

**Comment by kyriediculous** [comment]
Yeah there is a failing test indeed, not an assertion but a timing issue due to an afterEach hook and an async function call. They are passing from time to time however so it seems to be related to a timing issue. We're still investigating this. As you can see CI passed as the latest PRs were merged. 

This also seems to be specific to running with coverage, when running without coverage these tests pass just fine (at least from my experience so far).

File: hardhat.config.ts (hunk 1 above)
Comment (kyriediculous) [reply hardhat.config.ts]: Should probably add this to the eslint config !

**Review by yondonfu — APPROVED** [review]
LGTM!

**Pull request merged**